### PR TITLE
{AKS} `az aks update`: Add `--ip-families` to support updating cluster networking settings

### PR DESCRIPTION
**Pull request by tyler-lloyd**
---

This checklist is used to make sure that common guidelines for a pull request are followed.

### Related command

`az aks update --ip-families ipv4,ipv6`

### General Guidelines

- [x] Have you run `azdev style <YOUR_EXT>` locally? (`pip install azdev` required)
- [x] Have you run `python scripts/ci/test_index.py -q` locally? (`pip install wheel==0.30.0` required)

For new extensions:

- [x] My extension description/summary conforms to the [Extension Summary Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/extensions/extension_summary_guidelines.md).

### About Extension Publish

There is a pipeline to automatically build, upload and publish extension wheels.  
Once your pull request is merged into main branch, a new pull request will be created to update `src/index.json` automatically.  
You only need to update the version information in file setup.py and historical information in file HISTORY.rst in your PR but do not modify `src/index.json`.

### CODE REVIEW

**Comment by yonzhan** [comment]
AKS

**Review by FumingZhang — COMMENTED** [review]
The change generally looks good to me.

Could you please help add a scenario test (examples in file `test_aks_commands.py`), or update some of the existing ones to test the new option in `az aks update`?
- test_aks_create_dualstack_with_default_network
- test_aks_create_with_default_network
- test_aks_create_and_update_ipv6_count

**Comment by tyler-lloyd** [comment]
> The change generally looks good to me.
> 
> Could you please help add a scenario test (examples in file `test_aks_commands.py`), or update some of the existing ones to test the new option in `az aks update`?
> 
> * test_aks_create_dualstack_with_default_network
> * test_aks_create_with_default_network
> * test_aks_create_and_update_ipv6_count

@FumingZhang - the feature is not enabled yet and is only going to be tested by internal partners until we are ready for public preview. Can we agree to add a live test before we open to public preview?

**Review by FumingZhang — APPROVED** [review]
LGTM

**Review by yanzhudd — APPROVED** [review]

**Comment by tyler-lloyd** [comment]
@zhoxing-ms PTAL

File: src/aks-preview/azext_aks_preview/_params.py [hunk 1]
```diff
@@ -535,6 +535,7 @@ def load_arguments(self, _):
         c.argument('network_dataplane', arg_type=get_enum_type(network_dataplanes))
         c.argument('network_policy')
         c.argument('network_plugin', arg_type=get_enum_type(network_plugins))
+        c.argument('ip_families')
```
Comment (yanzhudd) [inline src/aks-preview/azext_aks_preview/_params.py:538]: If the value of this param is a list, please use `nargs='+'` to define it.
```suggestion
        c.argument('ip_families', nargs='+')
```

Comment (tyler-lloyd) [reply src/aks-preview/azext_aks_preview/_params.py:538]: @yanzhudd - would that be correct here? We treat the param as a string and then parse it here https://github.com/Azure/azure-cli-extensions/blob/850c41eb2fdea09fd30828a4047362a866c6bc93/src/aks-preview/azext_aks_preview/managed_cluster_decorator.py#L280C100-L280C100

Comment (yanzhudd) [reply src/aks-preview/azext_aks_preview/_params.py:538]: Hi @tyler-lloyd,
yeah, your logic works. But it is recommented to use `nargs='+'` to define parameters whose value is a list. After defining with that, the input will be automatically separated by spaces and you don't need to parse it manually.

Comment (tyler-lloyd) [reply src/aks-preview/azext_aks_preview/_params.py]: @yanzhudd - please review the last commit. is this what you're talking about? https://github.com/Azure/azure-cli-extensions/pull/7040/commits/d7ea54a3595a735b2ca8677ceb2b2609f7bc44f1

**Comment by yonzhan** [comment]
Please fix CI issues.

**Comment by tyler-lloyd** [comment]
@FumingZhang could you help take a look at the failing azdev style CI pipeline? None of the errors appear related to changes that I've made in this PR.

**Comment by FumingZhang** [comment]
image
Yep, I'll fix the lint issues in a separate PR later.

**Comment by tyler-lloyd** [comment]
> Please fix CI issues.

@yonzhan - CI issues are fixed now.

File: src/aks-preview/azext_aks_preview/_help.py [hunk 2]
```diff
@@ -869,6 +869,10 @@
         - name: --enable-disk-driver
           type: bool
           short-summary: Enable AzureDisk CSI Driver.
+        - name: --ip-families
+          type: string
+          short-summary: A comma separated list of IP versions to use for cluster networking.
```
Comment (yanzhudd) [inline src/aks-preview/azext_aks_preview/_help.py:874]: Please fix the short summary since it is space-separated now.

Comment (yanzhudd) [reply src/aks-preview/azext_aks_preview/_help.py:874]: That's the only comment I have for this PR, and everything else looks good to me.

Comment (tyler-lloyd) [reply src/aks-preview/azext_aks_preview/_help.py:874]: so does the user need to input `--ip-families ipv4 ipv6` and it won't support `--ip-families ipv4,ipv6` anymore? isn't that somewhat of a breaking change?

Comment (tyler-lloyd) [reply src/aks-preview/azext_aks_preview/_help.py:875]: bump @yanzhudd or @zhoxing-ms

Comment (tyler-lloyd) [reply src/aks-preview/azext_aks_preview/_help.py:875]: maybe @FumingZhang could answer?

Comment (FumingZhang) [reply src/aks-preview/azext_aks_preview/_help.py:875]: That sounds like a breaking behavior change, please confirm with PM if this is acceptable.

Comment (tyler-lloyd) [reply src/aks-preview/azext_aks_preview/_help.py:875]: we don't want the change if nargs means it has to be space separated. @yanzhudd please confirm

Comment (yanzhudd) [reply src/aks-preview/azext_aks_preview/_help.py:875]: Hi @tyler-lloyd,
sorry for the delay reply. I thought it was a newly added param and not a breaking change, hence the suggestion. If it is a breaking change, just keep it and do not change it to space separated, thanks!

**Comment by FumingZhang** [comment]
> image Yep, I'll fix the lint issues in a separate PR later.

I've fixed the style issues, please rebase from main to pull the fixes. And fix any error in the failed CI check *Azure.azure-cli-extensions (azdev style on Modified Extensions)*.

File: src/aks-preview/azext_aks_preview/_help.py (hunk 2 above)
Comment (tyler-lloyd) [reply src/aks-preview/azext_aks_preview/_help.py:875]: sounds good. reverted the nargs change and keeping it comma separated

**Review by yanzhudd — APPROVED** [review]

**Pull request merged**